src/main.cpp

```diff
@@ -1,19 +1,234 @@
+ *
+ * @return true if the mission is finished, false otherwise.
+ */
+bool WaypointNode::current_mission_finished() {
+    if (mission_progress >= 1.0) {
+        mission_progress = 0.0;
+        return true;
+    } else {
+        return false;
+    }
+}
+
+/**
+ * @brief Executes the event loop for the WaypointNode.
+ *
+ * This function is responsible for executing the event loop of the WaypointNode
+ * class. It checks the node state and performs the corresponding actions based
+ * on the state. The possible states are:
+ * - init: Initializes the mode.
+ * - pre_wait_time: Wait a predefined time.
+ * - reach_cruise_height: Fly drone to the cruising height.
+ * - fly_to_waypoint: Executes the mode to fly to the waypoint.
+ * - reach_target_height: Fly drone to the target height.
+ * - post_wait_time: Wait a predefined time.
+ *
+ * If the node state is unknown, an error message is logged and a job_finished
+ * error message is sent.
+ */
+void WaypointNode::event_loop() {
+    if (!this->get_active()) return;
+
+    switch (get_node_state()) {
+        case init:
+            mode_init();
+            break;
+        case pre_wait_time:
+            break;  // Wait for the `wait_timer` to trigger
+        case reach_cruise_height:
+            mode_reach_cruise_height();
+            break;
+        case fly_to_waypoint:
+            mode_fly_to_waypoint();
+            break;
+        case reach_target_height:
+            mode_reach_target_height();
+            break;
+        case post_wait_time:
+            break;  // Wait for the `wait_timer` to trigger
+        default:
+            RCLCPP_ERROR(this->get_logger(),
+                         "WaypointNode::%s: Unknown mission_state: %d",
+                         __func__, get_node_state());
+            this->job_finished("WaypointNode::" + (std::string) __func__ +
+                               ": Unknown mission_state");
+            reset_node();
+    }
+}
+
+/**
+ * @brief Resets the node for the next command.
+ *
+ * This function resets the state of the `WaypointNode` object for the next
+ * command. It sets the node state to `init`, sets the `state_first_loop` flag
+ * to `true`, resets the `mission_progress` to 0, and clears the `cmd` object.
+ */
+void WaypointNode::reset_node() {
+    set_node_state(init);
+    state_first_loop = true;
+    mission_progress = 0.0;
+    cmd = Command();
+}
+
+/**
+ * @brief Sets the node state to a new value.
+ *
+ * This function sets the node state to the specified new value. If the new
+ * value is different from the current node state, the `state_first_loop` flag
+ * is set to true. After updating the node state, a debug message is logged.
+ * Additionally, `mission_progress` is reset to 0.
+ *
+ * @param new_mission_state The new mission state to set.
+ */
+void WaypointNode::set_node_state(NodeState_t new_state) {
+    if (node_state != new_state) {
+        state_first_loop = true;
+    }
+
+    mission_progress = 0.0;
```

Comment:
Maybe add a variable for no and 100% progress

Reply:
I don't see a benefit